### PR TITLE
Add support for __all__ and potentilly deleting functions

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Adds support in modular for:
- deleting super()'s function (with raise Attribute Error)
- a way to call another `super`: if you inherit from both LlamaTokenier and PretrainedTokenizer, you can call `PretrainedTokenizer`'s `__init__` and it will be unravelled as `super().__init__()`
- splitting the __all__ that is created in the `modular_xxx.py`
- camel-cased renaming when dependencies are not found

### CODE REVIEW

File: src/transformers/models/gemma/modular_gemma.py [hunk 1]
```diff
@@ -164,6 +175,164 @@ def __init__(
...
+            pad_token=pad_token,
+            add_bos_token=add_bos_token,
+            add_eos_token=add_eos_token,
+            sp_model_kwargs=sp_model_kwargs,
+            clean_up_tokenization_spaces=clean_up_tokenization_spaces,
+            use_default_system_prompt=use_default_system_prompt,
+            spaces_between_special_tokens=spaces_between_special_tokens,
+            **kwargs,
+        )
+        del self.add_prefix_space
```
Comment (LysandreJik) [inline src/transformers/models/gemma/modular_gemma.py:267]: Can you add a comment above?

Comment (LysandreJik) [reply src/transformers/models/gemma/modular_gemma.py:267]: or on the same line

File: src/transformers/models/gemma/modular_gemma.py [hunk 2]
```diff
@@ -164,6 +175,164 @@ def __init__(
...
+        del self.legacy
+
+    def get_spm_processor(self):
+        raise AttributeError("Not needed for Gemma")
+
+    def unk_token_length(self):
+        raise AttributeError("Not needed for Gemma")
+
+    # skip
```
Comment (LysandreJik) [inline src/transformers/models/gemma/modular_gemma.py:276]: unclear, as seen with you in person

File: src/transformers/models/gemma/tokenization_gemma.py [hunk 3]
```diff
@@ -26,13 +30,14 @@
...
 if TYPE_CHECKING:
-    pass
+    from ...tokenization_utils_base import TextInput
 
 logger = logging.get_logger(__name__)
 
 VOCAB_FILES_NAMES = {"vocab_file": "tokenizer.model"}
 
 SPIECE_UNDERLINE = "▁"
+# fmt: on
```
Comment (LysandreJik) [inline src/transformers/models/gemma/tokenization_gemma.py:40]: :eyes:

File: src/transformers/models/gemma/tokenization_gemma.py [hunk 4]
```diff
@@ -110,96 +115,77 @@ def __init__(
         self.add_bos_token = add_bos_token
         self.add_eos_token = add_eos_token
         self.use_default_system_prompt = use_default_system_prompt
-
         self.sp_model = spm.SentencePieceProcessor(**self.sp_model_kwargs)
         self.sp_model.Load(vocab_file)
 
-        super().__init__(
+        PreTrainedTokenizer.__init__(
```
Comment (LysandreJik) [inline src/transformers/models/gemma/tokenization_gemma.py:121]: this should be a super no?

File: src/transformers/models/gemma/tokenization_gemma.py [hunk 5]
```diff
@@ -325,3 +307,35 @@ def create_token_type_ids_from_sequences(
...
+
+        if spaces_between_special_tokens:
+            sub_texts = " ".join(sub_texts)
+        else:
+            sub_texts = "".join(sub_texts)
+
+        return sub_texts.replace(SPIECE_UNDERLINE, " ")
+
+
+__all__ = ["GemmaTokenizer"]
```
Comment (LysandreJik) [inline src/transformers/models/gemma/tokenization_gemma.py:341]: clean!

**Review by LysandreJik — APPROVED** [review]
Looks good! Only left a few comments

File: utils/check_docstrings.py
Comment (ArthurZucker) [inline utils/check_docstrings.py:1]: this breaks some of the docstring updates

File: utils/modular_model_converter.py [hunk 6]
```diff
@@ -508,6 +522,36 @@ def replace_call_to_super(class_finder: ClassFinder, updated_node: cst.ClassDef,
...
+    # Remove the common suffix from the class_name
+    if common_suffix_len > 0:
+        new_part = class_name[:-common_suffix_len]
+    else:
+        new_part = class_name
+
+    # Convert the remaining new part to snake_case
+    snake_case = re.sub(r"(?<!^)(?=[A-Z])", "_", new_part).lower()
+
+    return snake_case
```
Comment (molbap) [inline utils/modular_model_converter.py:552]: That's very nice! It means new models will always have snake casing right?

Comment (ArthurZucker) [reply utils/modular_model_converter.py]: Should be yes!

**Pull request merged**